swift/apple/FirezoneNetworkExtension/Adapter.swift

```
@@ -258,119 +277,102 @@ class Adapter {
+      }
 
-      guard let data4 = routeListv4.data(using: .utf8),
-        let data6 = routeListv6.data(using: .utf8),
-        let decoded4 = try? JSONDecoder().decode([NetworkSettings.Cidr].self, from: data4),
-        let decoded6 = try? JSONDecoder().decode([NetworkSettings.Cidr].self, from: data6)
-      else {
-        fatalError("Could not decode route list from connlib")
+      // Apply network settings directly
+      guard let provider = packetTunnelProvider else {
+        Log.error(AdapterError.invalidSession(nil))
+        return
       }
 
-      let routes4 = decoded4.compactMap({ $0.asNEIPv4Route })
-      let routes6 = decoded6.compactMap({ $0.asNEIPv6Route })
+      Log.log("Setting interface config")
 
-      networkSettings.tunnelAddressIPv4 = tunnelAddressIPv4
-      networkSettings.tunnelAddressIPv6 = tunnelAddressIPv6
+      // Create network settings and apply configuration
+      let networkSettings = NetworkSettings(packetTunnelProvider: provider)
+      networkSettings.tunnelAddressIPv4 = ipv4
+      networkSettings.tunnelAddressIPv6 = ipv6
       networkSettings.dnsAddresses = dnsAddresses
-      networkSettings.routes4 = routes4
-      networkSettings.routes6 = routes6
       networkSettings.setSearchDomain(domain: searchDomain)
-      self.networkSettings = networkSettings
 
-      // Now that we have our interface configured, start listening for events. The first one will be us applying
-      // our network settings in the call below. We need the physical interface name macOS chooses for us in order
-      // to get the correct DNS resolvers on macOS. For that, we need the path parameter from the path update callback.
-      beginPathMonitoring()
+      // Parse and set IPv4 routes
+      if let routesData = ipv4Routes.data(using: .utf8),
+        let cidrs = try? JSONDecoder().decode([NetworkSettings.Cidr].self, from: routesData)
+      {
+        networkSettings.routes4 = cidrs.compactMap({ $0.asNEIPv4Route })
```

I think we should avoid making partial updates to the `networkSettings` here. Instead, I think we should try and decode everything first into local variables and then only if it all worked, try and apply it.

---

We can throw an exception here as long as we wrap the call-site in a `try catch`.

---

makes sense! will restore the previous behaviour